Recover pose from different cameras (version 2)

thezane opened the pull request:
Finish PR from #17981, which creates a new api for computing pose from different cameras.

The original author and commit is preserved.  I got the go-ahead from a maintainer to finish the original PR:  https://github.com/opencv/opencv/pull/17981#issuecomment-905237498

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake



```
force_builders_only=linux,docs
```

thezane commented on modules/calib3d/test/test_cameracalibration.cpp:
```diff
@@ -2158,19 +2161,24 @@ TEST(CV_RecoverPoseTest, regression_15341)
 
             // Estimation of fundamental matrix using the RANSAC algorithm
-            Mat E, E2, E3, R, t;
+            Mat E, E2, R, t;
```
I'm checking the different camera case first as the same camera case modifies the points input.

alalek commented on modules/calib3d/test/test_cameracalibration.cpp:
```diff
@@ -2142,6 +2142,10 @@ TEST(CV_RecoverPoseTest, regression_15341)
     // camera matrix with both focal lengths = 1, and principal point = (0, 0)
     const Mat cameraMatrix = Mat::eye(3, 3, CV_64F);
+
+    // camera matrix with focal lengths 0.5 and 0.6 respectively and principal point = (100, 200)
+    double cameraMatrix2Data[] = { 0.5, 0.6, 1, 0, 100, 200, 0, 0, 1 };
+    const Mat cameraMatrix2( 3, 3, CV_64F, cameraMatrix2Data );
```
Camera matrix has another data layout.

Check [documentation](https://docs.opencv.org/3.4/d9/d0c/group__calib3d.html) (camera intrinsic matrix A).

alalek commented on modules/calib3d/include/opencv2/calib3d.hpp:
```diff
@@ -2842,6 +2842,76 @@ unit length.
+@ref calibrateCamera. Furthermore, for the stereo case, additional flags may be used, see below.
+@param distCoeffs2 Input/output vector of distortion coefficients, the same as in
+@ref calibrateCamera.
+@param E The input essential matrix.
```
> E The input essential matrix.
> OutputArray E

input?
Please don't copy-paste!

alalek commented on modules/calib3d/src/five-point.cpp:
```diff
@@ -524,6 +524,31 @@ cv::Mat cv::findEssentialMat( InputArray points1, InputArray points2,
+    // Undistort image points, bring them to 3x3 identity "camera matrix"
+    Mat _pointsUntistorted1, _pointsUntistorted2;
+    undistortPoints(_points1, _pointsUntistorted1, cameraMatrix1, distCoeffs1);
+    undistortPoints(_points2, _pointsUntistorted2, cameraMatrix2, distCoeffs2);
```
findEssentialMat also calls undistortPoints internally.

If you want to provide new wrapper, it should be efficient.
Otherwise users can do all of that on their side using already existed API just using ten lines.

thezane replied:
Thanks for pointing this out.  I created a new private api for findEssentialMat, which takes undistorted points and two cameras.  The new findEssentialMat api is used by the existing findEssentialMat and new recoverPose apis.

Please let me know of any concerns with this approach.

alalek commented on modules/calib3d/test/test_cameracalibration.cpp:
```diff
@@ -2158,14 +2162,23 @@ TEST(CV_RecoverPoseTest, regression_15341)
+
+            // Check pose when camera matrices are different.
+            E = findEssentialMat(points1, points2, cameraMatrix, zeroDistCoeffs, cameraMatrix2, zeroDistCoeffs, RANSAC, 0.999, 1.0, mask);
+            Inliers = recoverPose(points1, points2, cameraMatrix, zeroDistCoeffs, cameraMatrix2, zeroDistCoeffs, E2, R, t, RANSAC, 0.999, 1.0, mask);
```
> zeroDistCoeffs

Non-zero distortion case should be tested too as new API suggests these parameters.

thezane commented:
Thanks for the review!  I addressed all comments except for unit tests for nonzero distortion.  I'll the tests this weekend.

thezane commented:
Thanks for the review!  I addressed all comments except for unit tests for nonzero distortion.  I'll add the tests this weekend.

thezane commented on modules/calib3d/src/five-point.cpp:
```diff
@@ -489,25 +512,10 @@ cv::Mat cv::findEssentialMat( InputArray _points1, InputArray _points2,
     CV_INSTRUMENT_REGION();
 
     // Undistort image points, bring them to 3x3 identity "camera matrix"
-    Mat _pointsUntistorted1, _pointsUntistorted2;
```
Undistorted is misspelled in original code.

thezane commented on modules/calib3d/src/five-point.cpp:
```diff
@@ -489,25 +512,10 @@ cv::Mat cv::findEssentialMat( InputArray _points1, InputArray _points2,
     // Undistort image points, bring them to 3x3 identity "camera matrix"
-    Mat _pointsUntistorted1, _pointsUntistorted2;
-    undistortPoints(_points1, _pointsUntistorted1, cameraMatrix1, distCoeffs1);
-    undistortPoints(_points2, _pointsUntistorted2, cameraMatrix2, distCoeffs2);
```
Undistorted is misspelled in original code.

thezane commented:
@alalek Thanks for your detailed review and appreciate your patience.

I addressed the comments.  Specifically, the latest changes ensure:
(1) The new recoverPose API only executes undistortPoints on each input points once.
(2) Existing API and functionality is preserved.

Please let me know of any other concerns.

thezane commented:
@alalek @asmorkalov  Gentle reminder about this PR.  It's been 2 weeks since I addressed concerns raised from the previous review.  Is there anything else that should be addressed?

asmorkalov reviewed (approved):
:+1: Well Done!

alalek commented on modules/calib3d/src/five-point.cpp:
```diff
@@ -401,6 +401,29 @@ class EMEstimatorCallback CV_FINAL : public PointSetRegistrator::Callback
 };
 
+// Find essential matrix given undistorted points and two cameras.
+static Mat findEssentialMat( InputArray _points1, InputArray _points2,
```
Consider adding underscore `_` suffix for internal names to avoid confusion with public API: `findEssentialMat_`

thezane commented on modules/calib3d/src/five-point.cpp:
```diff
@@ -401,6 +401,29 @@ class EMEstimatorCallback CV_FINAL : public PointSetRegistrator::Callback
+    Mat _pointsTransformed1, _pointsTransformed2;
+    Mat cm1 = cameraMatrix1.getMat(), cm2 = cameraMatrix2.getMat(), cm0;
+    Mat(cm1 + cm2).convertTo(cm0, CV_64F, 0.5);
+    CV_Assert(cm0.rows == 3 && cm0.cols == 3);
+    CV_Assert(std::abs(cm0.at<double>(2, 0)) < 1e-3 &&
+              std::abs(cm0.at<double>(2, 1)) < 1e-3 &&
+              std::abs(cm0.at<double>(2, 2) - 1.) < 1e-3);
+    Mat affine = cm0.rowRange(0, 2);
```
Calling `.rowRange()` is necessary as it allows the first 2 rows of `cm0` to be the 3rd argument of  `cv::transform`.  There, it's used to transform points in `_points1` and `_points2`.  The 3rd argument of `cv::transform` must be 2-by-2 or 2-by-3 as per [docs](https://docs.opencv.org/4.5.3/d2/de8/group__core__array.html#ga393164aa54bb9169ce0a8cc44e08ff22).

The logic in this function is not new to this PR. It originates from #17816.
